cli/src/semgrep/formatter/sarif.py

```diff
@@ -14,12 +16,109 @@
+
+        intermediate_vars_location_sarif = {
+            "location": {
+                "message": {"text": propagation_message_text},
+                "physicalLocation": {
+                    "artifactLocation": {"uri": str(rule_match.path)},
+                    "region": {
+                        "startLine": location.start.line,
+                        "startColumn": start_column,
+                        "endLine": location.end.line,
+                        "endColumn": end_column,
+                        "snippet": {"text": content},
+                        "message": {"text": propagation_message_text},
+                    },
+                },
+            }
+        }
+        return intermediate_vars_location_sarif
+
+    @staticmethod
+    def _rec_taint_obj_to_thread_flow_location_sarif(
+        var_type: str, taint_obj: Any, rule_match: RuleMatch
+    ) -> List[Any]:
+        taint_trace = []
+
+        if isinstance(taint_obj, out.CliMatchCallTrace):
+            taint_trace += SarifFormatter._rec_taint_obj_to_thread_flow_location_sarif(
+                var_type, taint_obj.value, rule_match
+            )
+
+        if isinstance(taint_obj, out.CliCall):
+            taint_trace += SarifFormatter._rec_taint_obj_to_thread_flow_location_sarif(
+                var_type, taint_obj.value[2], rule_match
+            )
+
+            for intermediate_var in taint_obj.value[1]:
+                taint_trace.append(
+                    SarifFormatter._taint_obj_intermediate_vars_to_thread_flow_location_sarif(
+                        intermediate_var, rule_match
+                    )
+                )
+
+        # the taint object can be an instance of CliLoc, CliClass or CliMatchCallTrace
+        if isinstance(taint_obj, out.CliLoc):
+            location = taint_obj.value[0]
+        elif isinstance(taint_obj, out.CliCall):
+            var_type = "Propagator"
+            location = taint_obj.value[0][0]
+        else:
+            return taint_trace
+
+        content, start_column, end_column = SarifFormatter._generate_taint_snippet(
+            location
+        )
+        source_message_text = f"{var_type} : '{content}' @ '{str(location.path)}:{str(location.start.line)}'"
+        location_sarif = {
+            "location": {
+                "message": {"text": source_message_text},
+                "physicalLocation": {
+                    "artifactLocation": {"uri": str(rule_match.path)},
+                    "region": {
+                        "startLine": location.start.line,
+                        "startColumn": start_column,
+                        "endLine": location.end.line,
+                        "endColumn": end_column,
+                        "snippet": {"text": content},
+                        "message": {"text": source_message_text},
+                    },
+                },
+            }
+        }
```

Comment:
If we define a function that takes the `var_type` and the `location` for returning that, it will avoid code duplication at several places

Reply:
Done